File: 01-Languages/Go/14-Async-and-Concurrency/main.go
// main.go - goroutines, channels, sync.WaitGroup (with real timing), select.
package main

import (
	"fmt"
	"sync"
	"time"
)

func worker(id int, results chan<- string) {
	results <- fmt.Sprintf("worker %d done", id)
}

func slowCompute(ms int, value int, results chan<- int) {
	time.Sleep(time.Duration(ms) * time.Millisecond)
	results <- value
}

func main() {
	fmt.Println("--- basic goroutine and channel ---")
	results := make(chan string)
	go worker(1, results)
	message := <-results
	fmt.Println(message)

	fmt.Println("\n--- sync.WaitGroup waiting for multiple goroutines ---")
	var wg sync.WaitGroup
	for i := 1; i <= 3; i++ {
		wg.Add(1)
		go func(id int) {
			defer wg.Done()
			fmt.Println("worker", id, "done (WaitGroup)")
		}(i)
	}
	wg.Wait()
	fmt.Println("All goroutines finished.")

	fmt.Println("\n--- sequential vs concurrent goroutines (real timing) ---")
	start1 := time.Now()
	seqResults := make(chan int, 3)
	slowCompute(80, 1, seqResults)
	slowCompute(80, 2, seqResults)
	slowCompute(80, 3, seqResults)
	fmt.Println("Sequential 3x80ms calls took ~", time.Since(start1).Milliseconds(), "ms")

	start2 := time.Now()
	concResults := make(chan int, 3)
	go slowCompute(80, 1, concResults)
	go slowCompute(80, 2, concResults)
	go slowCompute(80, 3, concResults)
	total := 0
	for i := 0; i < 3; i++ {
		total += <-concResults
	}
	fmt.Println("Concurrent goroutines of the same 3x80ms tasks took ~", time.Since(start2).Milliseconds(), "ms, total=", total)

	fmt.Println("\n--- select: whichever channel is ready first ---")
	channelA := make(chan string)
	channelB := make(chan string)
	go func() {
		time.Sleep(30 * time.Millisecond)
		channelA <- "from A"
	}()
	go func() {
		time.Sleep(60 * time.Millisecond)
		channelB <- "from B"
	}()
	select {
	case msg := <-channelA:
		fmt.Println("select got:", msg)
	case msg := <-channelB:
		fmt.Println("select got:", msg)
	}
}

File: 01-Languages/Rust/14-Async-and-Concurrency/main.rs
// main.rs - std::thread::spawn (with real timing), Arc<Mutex<T>> for safe shared mutable state.

use std::sync::{Arc, Mutex};
use std::thread;
use std::time::{Duration, Instant};

fn slow_compute(ms: u64, value: i32) -> i32 {
    thread::sleep(Duration::from_millis(ms));
    value
}

fn main() {
    println!("--- basic spawned thread ---");
    let handle = thread::spawn(|| {
        println!("Hello from a thread");
    });
    handle.join().unwrap();

    println!("\n--- sequential vs concurrent threads (real timing) ---");
    let start1 = Instant::now();
    let a = slow_compute(80, 1);
    let b = slow_compute(80, 2);
    let c = slow_compute(80, 3);
    println!(
        "Sequential 3x80ms calls took ~{}ms (sum={})",
        start1.elapsed().as_millis(),
        a + b + c
    );

    let start2 = Instant::now();
    let h1 = thread::spawn(|| slow_compute(80, 1));
    let h2 = thread::spawn(|| slow_compute(80, 2));
    let h3 = thread::spawn(|| slow_compute(80, 3));
    let total = h1.join().unwrap() + h2.join().unwrap() + h3.join().unwrap();
    println!(
        "Concurrent threads of the same 3x80ms tasks took ~{}ms (sum={})",
        start2.elapsed().as_millis(),
        total
    );

    println!("\n--- Arc<Mutex<T>> for safe shared mutable state ---");
    let counter = Arc::new(Mutex::new(0));
    let mut handles = vec![];

    for _ in 0..10 {
        let counter = Arc::clone(&counter);
        let handle = thread::spawn(move || {
            let mut num = counter.lock().unwrap();
            *num += 1;
        });
        handles.push(handle);
    }

    for handle in handles {
        handle.join().unwrap();
    }

    println!("Result (reliably 10, compiler-enforced): {}", *counter.lock().unwrap());
}
